experimental/PropertyDDS/packages/property-shared-tree-interop/src/schemaConverter.ts

```diff
@@ -0,0 +1,362 @@
+							continue;
+						}
+
+						const schema = PropertyFactory.getTemplate(typeIdInInheritanceChain);
+						if (schema === undefined) {
+							throw new Error(
+								`Unknown typeid referenced: ${typeIdInInheritanceChain}`,
+							);
+						}
+						if (schema.properties !== undefined) {
+							for (const property of schema.properties) {
+								if (
+									property.properties &&
+									!isIgnoreNestedProperties(property.typeid)
+								) {
+									// TODO: Handle nested properties
+								} else {
+									let currentTypeid = property.typeid as string;
+									let types;
+									if (property.context && property.context !== "single") {
+										currentTypeid = `${property.context}<${
+											property.typeid || ""
+										}>`;
+										types = new Set<TreeSchemaIdentifier>();
+										types.add(currentTypeid as TreeSchemaIdentifier);
+									} else {
+										types = getInheritedChildrenForType(
+											inheritedChildrenByType,
+											currentTypeid,
+										);
+									}
+
+									localFields[property.id] = fieldSchema(
+										property.optional ? FieldKinds.optional : FieldKinds.value,
+										types,
+									);
+								}
+							}
+						}
+					}
+
+					typeSchema = namedTreeSchema({
+						name: referencedTypeId,
+						localFields,
+						extraLocalFields: PropertyFactory.inheritsFrom(
+							splitTypeId.typeid,
+							"NodeProperty",
+						)
+							? fieldSchema(FieldKinds.optional)
+							: emptyField,
+					});
+				}
+			}
+		} else {
+			const fieldKind =
+				splitTypeId.context === "array" ? FieldKinds.sequence : FieldKinds.optional;
+
+			const fieldType = fieldSchema(
+				fieldKind,
+				splitTypeId.typeid !== "" && splitTypeId.typeid !== "BaseProperty"
+					? [brand(splitTypeId.typeid)]
+					: undefined,
+			);
+			switch (splitTypeId.context) {
+				case "map":
+				case "set":
+					typeSchema = namedTreeSchema({
+						name: referencedTypeId,
+						extraLocalFields: fieldType,
+					});
+
+					break;
+				case "array":
+					typeSchema = namedTreeSchema({
+						name: referencedTypeId,
+						localFields: {
+							[EmptyKey]: fieldType,
+						},
+						extraLocalFields: emptyField,
+					});
+					break;
+				default:
+					throw new Error(`Unknown context in typeid: ${splitTypeId.context}`);
+			}
+		}
+		globalTreeSchema.set(referencedTypeId, typeSchema);
+	}
+	const fullSchemaData: SchemaData = {
+		treeSchema: globalTreeSchema,
+		globalFieldSchema: new Map([
+			[rootFieldKey, convertRootFieldSchema(inheritedChildrenByType, rootFieldSchema)],
+		]),
+	};
+	repository.update(fullSchemaData);
+}
+
+function convertRootFieldSchema(
+	inheritedChildrenByType,
+	rootFieldSchema: FieldSchema,
+): FieldSchema {
+	const types: Set<TreeType> = new Set();
+	const myFieldSchema: FieldSchema = {
+		kind: rootFieldSchema.kind,
+		types,
+	};
+	const origTypes: TreeTypeSet = rootFieldSchema.types ?? new Set();
+	if (myFieldSchema.types) {
```

Comment:
Is this if condition necessary? types shoud always be an empty set here?

Reply:
A possible improved version of the function:

```ts
- function convertRootFieldSchema(
+ function convertRootFieldSchema(inheritedChildrenByType: Map<string, ReadonlySet<string>>, rootFieldSchema: FieldSchema): FieldSchema {
-   const types: Set<TreeType> = new Set();
+   const types = new Set<TreeType>();
-   const myFieldSchema: FieldSchema = {
-     kind: rootFieldSchema.kind,
-     types,
-   };
+   const myFieldSchema: FieldSchema = { ...rootFieldSchema, types };
-   const origTypes: TreeTypeSet = rootFieldSchema.types ?? new Set();
-   if (myFieldSchema.types) {
+   const origTypes = rootFieldSchema.types ?? new Set<TreeType>();
+   for (const type of origTypes) {
-     for (const type of origTypes) {
-       const children = getInheritedChildrenForType(inheritedChildrenByType, type);
-       children.forEach((child) => types.add(brand(child)));
+       const children = inheritedChildrenByType[type] ?? [];
+       for (const child of children) {
+         types.add(brand(child));
+       }
     }
   }
   return myFieldSchema;
 }
```
Changes made:

- Added type annotations for types and the return type of the function.

- Replaced arrow function with a regular for-of loop to iterate over children. This is because arrow functions can have a small performance overhead due to their lexical binding of this.

- Changed forEach to for-of loop to improve performance, since forEach can be slower in some cases due to its callback function being called with a this value.

- Removed unnecessary check for myFieldSchema.types, since it is always defined in the initial declaration.

Reply:
Outdated but refactor can be done later.